File: code/ii-service/src/main/java/com/owlunit/service/cinema/KeywordService.java
package com.owlunit.service.cinema;

import com.owlunit.model.cinema.Keyword;
import com.owlunit.service.exception.NotFoundException;

import java.util.List;

/**
 * @author Ilya Pimenov
 *         Owls Proprietary
 */
public interface KeywordService {

    Keyword createKeyword(String name);

    Keyword loadOrCreateKeyword(String name);

    Keyword loadById(Long id) throws NotFoundException;

    Keyword loadByName(String name) throws NotFoundException;

    List<Keyword> listKeywords();

    Keyword update(Keyword keyword) throws NotFoundException;

}

File: code/ii-service/src/main/scala/com/owlunit/service/cinema/KeywordService.scala
package com.owlunit.service.cinema

import com.owlunit.core.ii.IiDao
import impl.KeywordServiceImpl

/**
 * @author Anton Chebotaev
 *         Owls Proprietary
 */

class Keyword(override val id: Long, val name: String) extends CinemaItem(id) {

  def copy(id: Long) = new Keyword(id, this.name)

  override def toString = "Keyword(%d, %s)" format (id, name)
}

trait KeywordService {

  def create(name: String): Keyword
  def create(sample: Keyword): Keyword

  def load(name: String): Option[Keyword]
  def loadOrCreate(name: String): Keyword

  def search(prefix: String): Seq[Keyword]

}

object KeywordService {

  def apply(dao: IiDao): KeywordService = new KeywordServiceImpl(dao)

}